src/ORM/Behavior/TreeBehavior.php

```diff
@@ -533,20 +524,24 @@ public function findTreeList(SelectQuery $query, array $options): SelectQuery
      * @param array<string, mixed> $options Array of options as described above.
      * @return \Cake\ORM\Query\SelectQuery Augmented query.
      */
-    public function formatTreeList(SelectQuery $query, array $options = []): SelectQuery
-    {
-        return $query->formatResults(function (CollectionInterface $results) use ($options) {
-            $options += [
-                'keyPath' => $this->_getPrimaryKey(),
-                'valuePath' => $this->_table->getDisplayField(),
-                'spacer' => '_',
-            ];
-
-            $nested = $results->listNested();
-            assert($nested instanceof TreeIterator);
-
-            return $nested->printer($options['valuePath'], $options['keyPath'], $options['spacer']);
-        });
+    public function formatTreeList(
+        SelectQuery $query,
+        Closure|string|null $keyPath = null,
+        Closure|string|null $valuePath = null,
+        ?string $spacer = null
```

Comment:
Love this. It gives us better autocomplete, and a simpler API that allows not null constraints to be enforced by the type system :clap:

I don't know how we make this an easy to do upgrade though. We might be able to use rector if we write some custom logic.

Reply:
Yup,  for e.g. for `findChildren()` the arg `$for` is a required and we no longer need to check that `$options['for']` is set.